Various doc updates

gvanrossum opened the pull request.

gvanrossum commented:
@sixolet Can you review the docs I wrote for flexible callable here?

sixolet commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+
+As an experimental mypy extension, you can specify ``Callable`` types
+that support keyword arguments, optional arguments, and more.  The
+"argument" part of the ``Callable`` specification (the part specifying
```
Where you specify the arguments of a ``Callable``, you can choose to supply just the type of a nameless positional argument, or an "argument specifier" representing a more complicated form of argument.

sixolet commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+            **kwargs: int) -> int:
+       ...
+
+   F = Callable[[Arg(int),  # No argument name given
```
Consider making this just `int`, to show that a raw type has no arg name and that you can mix them?

sixolet commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+
+   MyFunc = Callable[[Arg(int), Arg(str), Arg(int)], float]
+
+A ``Callable`` with unspecified argument types, such as
```
This equivalency is near, but not entirely true.  A `Callable[..., int]` gets `is_ellipsis_args` set to `True`, unlike `Callable[[VarArg(), KwArg()], int]`.  A Callable with ellipsis args is not only considered a subtype of all Callables with compatible return types (which it naturally is), but also a supertype of all Callables with compatible return types (which is weird and special).  This is the only other wart in the type system I know of that's comparable to `Any`.

sixolet commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+.. note::
+
+   This feature is experimental.  Details of the implementation may
+   change and there may be unknown limitations. **IMPORTANT:** At
```
Remove the first "at runtime" and maybe make it "Each of the functions above currently returns its ``type`` argument..." 

(to remove the awkwardly repeated phrase)

JelleZijlstra commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+                 Arg(int, 'b'),
+                 DefaultArg(int, 'c'),
+                 VarArg(int),
+                 NamedArg(int, 'd'),
```
This is probably the wrong place for this discussion, but why don't we call this `KeywordOnlyArg`? `NamedArg` is confusing because normal args are also named.

gvanrossum replied:
That discussion can go in https://github.com/python/typing/issues/264, which itself has a few naming differences from Naomi's PRs (#2426 and https://github.com/python/typeshed/issues/684). Or it can go into Naomi's PEP once she's written it.

JelleZijlstra commented on docs/source/python36.rst:
```diff
@@ -16,6 +16,7 @@ now have type annotations using either of the two forms:
 
 .. code-block:: python
 
+   from typing import Optionsl
```
Optional

gvanrossum commented:
Addressed the review comments. I decided not to explain in detail the difference between Callable[..., T] and the other form -- it's esoteric and hard to explain and nobody outside mypy dev circles cares.

sixolet commented on docs/source/kinds_of_types.rst:
```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
+
+   MyOtherFunc = Callable[..., int]
+
+is (roughly) equivalent to
```
I like "roughly" here 😈
